### PR TITLE
Use communicate in shell task

### PR DESCRIPTION
**Pull request by eapolinario**
# TL;DR
Use a different method to wait for child task in a ShellTask

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
The current implementation of `ShellTask` might hit the deadlock described in https://docs.python.org/3/library/subprocess.html#subprocess.Popen.wait if the child process ends up writing more than the pipe buffer size (which in most Linux systems sits around 65KB, or 16 pages).

The fix is to use [Popen.communicate](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate), which waits for the process to terminate and returns a tuple `(stdout_data, stderr_data)`. The one downside of this is that it keeps both stdout and stderr in memory. We might expose a flag to use the slower method of using temporary files to hold the child process's stdout and stderr in the future.

The test I added demonstrates the deadlock with the old code.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/4313

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

**Review by kumare3 — DISMISSED** [review]

**Comment by eapolinario** [comment]
The test I added is failing on windows with this error message:
```
gen_script = '#!/bin/bash&&&&set -uex&&&&for i in $(seq 1 200000)&&do&&  echo "This is an error message" >&2&&done&&&&echo "This is the output of the program"'
```

This is the script I'm trying to run, but why are those '&' being added there?

**Review by wild-endeavor — APPROVED** [review]

**Review by pingsutw — APPROVED** [review]
